src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -1017,7 +1022,18 @@ public override void Write(byte[] buffer, int offset, int count)
                     int bytesRead = _deflateStream._inflater.Inflate(new Span<byte>(_arrayPoolBuffer));
                     if (bytesRead > 0)
                     {
+                        bool greaterThanUncompressedSize = false;
+                        if (_deflateStream._uncompressedSize > 0 && _deflateStream._uncompressedSize < bytesRead)
+                        {
+                            bytesRead = (int)_deflateStream._uncompressedSize;
+                            greaterThanUncompressedSize = true;
+                        }
+                        else
+                            _deflateStream._uncompressedSize -= bytesRead;
+
                         _destination.Write(_arrayPoolBuffer, 0, bytesRead);
+                        if (greaterThanUncompressedSize)
+                            throw new InvalidDataException(SR.GenericInvalidData);
```

Comment:
> I thought we had discussed that there was precedent for truncation since that's what's already happening with compressedSize. 

Yeah we talked about it, but that time i haven't had checked this behavior with other app and after that I feel throwing is better ...
> It's a bit unusual to throw and have a contract that the object is still safe to interact with

The issue reported that throwing is expected, when tried with WinZip, 7zip I got corrupted error messages and only _ucompressedSize section of file was decompressed, well WinZip was asking if I would like to continue unzipping entire file even if it is corrupted. Anyways, that is an option for Apps using this API. But I believe the Apps should know if there were any invalid/corrupted data and choose how to proceed. So preferred to throw, of course open for more discussion

> but I suppose if you document it and test all the edge cases for well defined behavior it might be ok.

All existing tests passed with this change, for now i have only test for this scenario which is passed, might create more but that leads to another question i wanted to discuss: the existing test files not enough to create more meaningful tests, i tried to add new test file but even uploaded to nuget with new version, but some of existing tests failing with new nuget version (some byte difference in Assert), so reverted back that change

Reply:
> The issue reported that throwing is expected

The reported issue is referring to an inconsistency between the local header and central header, for that we should throw.

Did you actually test the scenario where the headers were consistent but the payload was too long?  It's not clear to me if we have sufficient precedent to throw in that case.

> All existing tests passed with this change

That's not what I'm suggesting.  I'm saying that if you document that it is "ok" to handle this exception and still interact with the zip archive and entry, then you really need to test that the state of the objects is still consistent after handling an exception like this.   To test that you'd have to write new tests that catch the thrown exception then fully exercise all relevant functionality.

Reply:
> Did you actually test the scenario where the headers were consistent but the payload was too long? It's not clear to me if we have sufficient precedent to throw in that case.

Yes I did, but for now have only one test for that https://github.com/buyaa-n/corefx/pull/2/files#diff-ee82cb4b115f77168353a3d82a87037cR154

Reply:
> That's not what I'm suggesting. I'm saying that if you document that it is "ok" to handle this exception and still interact with the zip archive and entry, then you really need to test that the state of the objects is still consistent after handling an exception like this. To test that you'd have to write new tests that catch the thrown exception then fully exercise all relevant functionality.

The only test i have only checking the copied stream size after catching the exception, by `then fully exercise all relevant functionality` you mean if the received stream and/or ZAEntry function properly? sure I can add such test

Reply:
I'm suggesting the implications of such an assertion, not necessarily telling you you have to do that.  *If* you say that it's safe to catch that exception and continue using the object then you better make sure all the functionality of the object still works.  I'm trying to illustrate why that's a hard thing to say, since you need to test everything.  These objects are stateful objects, so it can be very tricky to make a claim like its safe to catch an exception and continue using it.  I think you'd be better off either saying you cannot use the object after it throw, or just truncate the data (like we truncate the compressed data).

Reply:
Sorry still not getting the exact point, but I assume you mean if the read stream will be fully functional, as you know there is 2 read mode, Read only and Update:
1. For read only mode all functionalities should be work
2. For update mode I assume it should work, but yes I need to test that and I will add more test for that

Reply:
It's more like: if you tell someone it is OK to catch an exception and still use the stream, are you garunteeing that the stream is in a safe state?  Here's a list of the sort of things I'd expect to work:
1. Position represents a legitimate value and not some value greater than the size you were trying to limit.
2. Subsequent reads either return nothing or will throw in the same way.
3. Seeks on the stream do a reasonable thing (either throw, or behave as if seeking from the limited value).
4. many others

Essentially, you need to be *very careful* that the state the stream remains in provides legitimate behavior if you're telling folks its safe to catch an exception.

I strongly believe that you're better off implmenting this as truncation first, so that you get all the behaviors correct.  Then you can decide if you want to throw when it truncates.

Reply:
> Essentially, you need to be very careful that the state the stream remains in provides legitimate behavior if you're telling folks its safe to catch an exception.
I strongly believe that you're better off implementing this as truncation first, so that you get all the behaviors correct. Then you can decide if you want to throw when it truncates.

Yes truncating the output stream fist and throwing after that, as deflate stream is closed in finally block base deflate stream is not readable anymore. 

> Here's a list of the sort of things I'd expect to work:
> 1. Position represents a legitimate value and not some value greater than the size you were trying to limit.
> 2. Subsequent reads either return nothing or will throw in the same way.
> 3. Seeks on the stream do a reasonable thing (either throw, or behave as if seeking from the limited value).
> 4. many others

Added assertions for base deflate stream, will add more test for output read stream is working as expected. Also will add tests for update mode

Reply:
Have got issue with update mode. in this mode when `InvalidDataException` occurs it nulls output stream https://github.com/buyaa-n/corefx/pull/2/files#diff-ba8d5a9ef0ce8bd24cf0576f0ab0972aL367
Now i don't know what to do:

1.  allow reading stream after throw only for `Read` mode and leave Update mode as is. 
2. or don't throw and return only truncated stream, in that case i assume base compressed stream remain open/readable which will be no use of this fix
3. or just throw and don't truncate read value, in that case as inflate and read/write operation is done in chunks user will still received truncated stream up to before throw
4. or maybe throw any other exception than `InvalidDataException` and let both modes use truncated stream, in that case Update mode need to be tested thoroughly

This is blocking me to raise PR in corefx, or maybe raise PR hear others opinion?

Reply:
After thinking about this little more i think the solution number 1 or 4 is more meaningful and what i have on this PR is solution 1, so am going to raise PR on coreFX and see others feedback

Reply:
Having Read throw and Update only truncate seems reasonable.  Not sure if that's what you meant by 1.  You could implement with an additional optional property passed to the inflater, just catch the exception, or limit the bytes copied out to the memory stream during the read portion of Update.

Reply:
Option 1 is what I have now, throw `InvalidDataException` and return truncated stream, in case of Read mode stream will be readable up to uncompressed size, but in case of Update mode it catching the exception and setting returning truncated steam to null. With that i mean we throw for both case but  in update mode returning stream will not be readable.

The solution you are suggesting is closer to option 4. Where we could throw with something else other than `InvalidDataException` so that in update mode the returning stream could be readable. We need to throw both cases to close the underlying deflate stream and make it not readable more. Actually there is no separate read for Update mode, as it reads the compressed stream using read mode and wraps the returned stream with seekable one

Reply:
I understand that but you can always make the behavior different if you want since you are creating the streams and are able to pass state in.

I really think that you can solve all of these issues if you just truncate the data.  The library is **already** truncating if compressed size is smaller than the compressed payload, the fact that you are throwing when the uncompressed size is smaller than compressed size, and still trying to make the stream be in a usable state after catching the assembly is inconsistent with existing behavior and proving to be difficult as I predicted.

Reply:
> I really think that you can solve all of these issues if you just truncate the data

I didn't found a way to stop decompress without throwing, but I have got one idea might work, will give it try.

>  the fact that you are throwing when the uncompressed size is smaller than compressed size

just a little correction: throwing when decompressed amount goes over uncomressed size